blockchain/store_test.go

```diff
@@ -0,0 +1,263 @@
+	})}
+	validPartSet := types.NewPartSetFromHeader(types.PartSetHeader{Total: 2})
+	validPartSet.AddPart(part1, false)
+	validPartSet.AddPart(part2, false)
+
+	incompletePartSet := types.NewPartSetFromHeader(types.PartSetHeader{Total: 2})
+
+	uncontiguousPartSet := types.NewPartSetFromHeader(types.PartSetHeader{Total: 0})
+	uncontiguousPartSet.AddPart(part2, false)
+	seenCommit1 := &types.Commit{Precommits: []*types.Vote{{Height: 10}}}
+
+	// End of setup, test data
+
+	tuples := []struct {
+		block            *types.Block
+		parts            *types.PartSet
+		seenCommit       *types.Commit
+		wantErr          bool
+		wantPanic        bool
+		eraseDBForCommit bool
+		skipReason       string
+	}{
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:      validPartSet,
+			seenCommit: seenCommit1,
+		},
+
+		{
+			block:     nil,
+			wantPanic: true, // want a panic on saving nil block
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  4,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:     uncontiguousPartSet,
+			wantPanic: true, // and incomplete and uncontiguous parts
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:     incompletePartSet,
+			wantPanic: true, // incomplete parts
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:            validPartSet,
+			seenCommit:       seenCommit1,
+			eraseDBForCommit: true,
+			skipReason:       "Waiting on https://github.com/tendermint/tmlibs/pull/56 to be merged first",
+			wantPanic:        true, // erase all content so that the commit will not be read
+		},
+	}
+
+	type quad struct {
+		block  *types.Block
+		commit *types.Commit
+		meta   *types.BlockMeta
+
+		seenCommit *types.Commit
+	}
+
+	for i, tuple := range tuples {
+		if tuple.skipReason != "" {
+			t.Logf("#%d: skipping this test case because of reason:\n%s", i, tuple.skipReason)
+			continue
+		}
+		bs, db := freshBlockStore()
+		// SaveBlock
+		res, err, panicErr := doFn(func() (interface{}, error) {
+			bs.SaveBlock(tuple.block, tuple.parts, tuple.seenCommit)
+			if tuple.block == nil {
+				return nil, nil
+			}
+			bBlock := bs.LoadBlock(tuple.block.Height)
+			bBlockMeta := bs.LoadBlockMeta(tuple.block.Height)
+			bSeenCommit := bs.LoadSeenCommit(tuple.block.Height)
+			if tuple.eraseDBForCommit {
```

Comment:
replayed in the associated PR. I think we should not erase db by calling `Close()`. feels wrong

Reply:
Can you reattach to the in-mem db after closing it?

Reply:
@melekes thank you for the advocacy, I've reflected on your comments in the PR https://github.com/tendermint/tmlibs/pull/56 and I concede with your thoughts :)

@adrianbrink yes we can :) However, I'll instead invoke an iterator and delete all the content in the DB for this test instead of invoking .Close()

Reply:
I've updated the code, no longer need to invoke db.Close(), I now have a helper that literally erases the DB, PTAL @melekes and @adrianbrink, thanks.